examples/ex25p.cpp

```diff
@@ -0,0 +1,1050 @@
+//               propagation problem corresponding to the second order indefinite
+//               Maxwell equation (1/mu) * curl curl E - \omega^2 * epsilon E = f
+//               with a Perfectly Matched Layer (PML).
+//               We discretize with Nedelec finite elements in 2D or 3D.
+//
+//               The example also demonstrates the use of complex valued bilinear and linear forms.
+//               We recommend viewing example 22 before viewing this example.
+//               Four examples are provided with exact solutions (iprob = 0-3).
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+#ifdef _WIN32
+#define jn(n, x) _jn(n, x)
+#define yn(n, x) _yn(n, x)
+#endif
+
+using namespace std;
+using namespace mfem;
+
+// Class for setting up a simple Cartesian PML region
+class CartesianPML
+{
+private:
+   Mesh *mesh;
+   int dim;
+
+   // Length of the PML Region in each direction
+   Array2D<double> length;
+
+   // Computational Domain Boundary
+   Array2D<double> comp_dom_bdr;
+
+   // Domain Boundary
+   Array2D<double> dom_bdr;
+
+   // Integer Array identifying elements in the pml
+   // 0: in the pml, 1: not in the pml
+   Array<int> elems;
+
+   // Method computing PML data
+   void SetBoundaries();
+
+public:
+   // Constructor
+   CartesianPML(Mesh *mesh_,Array2D<double> length_);
+
+   // Return Computational Domain Boundary
+   Array2D<double> GetCompDomainBdr() {return comp_dom_bdr;}
+
+   // Return Domain Boundary
+   Array2D<double> GetDomainBdr() {return dom_bdr;}
+
+   // Return Marker list for elements
+   Array<int> * GetMarkedPMLElements() {return &elems;}
+
+   // Mark element in the PML region
+   void SetAttributes(ParMesh *pmesh);
+
+   // PML complex stretching function
+   void StretchFunction(const Vector &x, vector<complex<double>> &dxs);
+};
+
+class PmlMatrixCoefficient : public MatrixCoefficient
+{
+private:
+   CartesianPML * pml = nullptr;
+   void (*Function)(const Vector &, CartesianPML * , DenseMatrix &);
+public:
+   PmlMatrixCoefficient(int dim, void(*F)(const Vector &, CartesianPML *,
+                                          DenseMatrix &),
+                        CartesianPML * pml_)
+      : MatrixCoefficient(dim), pml(pml_), Function(F)
+   {}
+   virtual void Eval(DenseMatrix &K, ElementTransformation &T,
+                     const IntegrationPoint &ip)
+   {
+      double x[3];
+      Vector transip(x, 3);
+      T.Transform(ip, transip);
+      K.SetSize(height, width);
+      (*Function)(transip, pml, K);
+   }
+};
+
+void maxwell_solution(const Vector &x, vector<complex<double>> &Eval);
+
+void E_bdr_data_Re(const Vector &x, Vector &E);
+void E_bdr_data_Im(const Vector &x, Vector &E);
+
+void E_exact_Re(const Vector &x, Vector &E);
+void E_exact_Im(const Vector &x, Vector &E);
+
+void source(const Vector &x, Vector & f);
+
+void detJ_JT_J_inv_Re(const Vector &x, CartesianPML * pml, DenseMatrix &M);
```

Comment:
It's not clear what `J` refers to.  I think I know what you mean but some explanation would be useful.  Perhaps the better place to document this is above the `PmlMatrixCoefficient`.  We could get away with a brief explanation of the concept along with a reference to a standard publication.

Reply:
Added brief explanation